### DISCUSSION TITLE
[RFC] Design Systems + Theming

### DISCUSSION THREAD

**Discussion by thescientist13** [ProjectEvergreen/greenwood#800] [General]
## Type of Change
- [x] New Feature Request
- [ ] Documentation / Website
- [ ] Improvement / Suggestion
- [ ] Bug
- [ ] Other (please clarify below)

## Summary
Part of maintaining any website with some degree of UX consistency tends to require a bit of planning and implementation around a design system.  A way of expressing and reusing styles across a website and its components, from things like fonts and icons, to color, to heading styles, and all sorts of readability and formatting related considerations that make any website a pleasure to use.

This RFC aims to start the discussion around how to manage these sorts of elements and how to implement them in a Web Components world, mostly when dealing with Shadow DOM.

## Details
To dig into the details a bit more, some technologies / topics that should be evaluated and explored are:
1. Usage (pros / cons) of ShadowDOM vs LightDOM, and interop with features like CSS variables, nested selectors
1. Theming in general, like even just making some examples or boilerplates (Greenwood templates / themes could ultimately be its own RFC)
1. How technologies like [CSS Modules](https://github.com/css-modules/css-modules) and [Styled Components](https://www.styled-components.com/) fit into the story
1. Documentation vs implementation, e.g. what should be in greenwood itself vs what should be supported in a way the developers can (ideally) pick based on their own preferences?
1. How does this all fit into the publishing (to npm) story?  Some [context](https://github.com/ProjectEvergreen/greenwood/pull/108#issuecomment-506985111).
1. Are preprocessors like SASS / LESS still needed, or is PostCSS enough?

I'm sure there's more, but this should get the conversation going and we can use the website to create POC branches to try different things out.  Ideally avoid picking winners / losers and hopefully being able to support multiple workflows instead.  We might pick our own for the website, but it shouldn't necessarily be defacto for greenwood _specifically_, if that makes sense.

Related Articles
- https://css-tricks.com/styling-a-web-component/
- https://developers.google.com/web/fundamentals/web-components/shadowdom
- https://www.smashingmagazine.com/2016/12/styling-web-components-using-a-shared-style-sheet/

**Comment by thescientist13** [comment]
Gatsby Themes
https://www.gatsbyjs.org/blog/2019-07-03-customizing-styles-in-gatsby-themes-with-theme-ui/

**Comment by aholtzman** [comment]
My fist impression of Gatsby Themes is that it seems like an awful lot of complexity for a limited number of cases where it would truly be useful. Having CSS-in-JS as part of the mix makes sense, but I think our need should be served with something leaner and more open. I found this short article about using Emotion with Web Components - seems like a promising path to pursue.

https://css-tricks.com/web-standards-meet-user-land-using-css-in-js-to-style-custom-elements/

I played around a bit with Styled-Components in the site build but with no success; I have not used Emotion yet but it looks easy enough and it not dependent on JSX.

**Comment by thescientist13** [comment]
Should also have some documentation practices around handling [FOUC](https://en.wikipedia.org/wiki/Flash_of_unstyled_content) and things like [loading fonts](https://github.com/ProjectEvergreen/greenwood/pull/164) from the build and [not from Google](https://github.com/ProvidenceGeeks/website-frontend/issues/152).

**Comment by thescientist13** [comment]
TIL `LitElement` recommends using the `static` styles property
https://lit-element.polymer-project.org/guide/styles#static-styles

Also be aware of `unsafeCss`
https://lit-element.polymer-project.org/guide/styles#static-styles

example usage, works nice with external files too + `:host`!
https://github.com/thegreenhouseio/www.thegreenhouse.io/pull/123

**Comment by thescientist13** [comment]
Something worth documenting somewhere, or a good trick in general, but if you want to disable ShadowDOM for a LitElement, you can use
```js
createRenderRoot() {
  return this;
```

**Comment by thescientist13** [comment]
An article on design systems when moving from stencil to lit
https://www.abeautifulsite.net/posts/moving-from-stencil-to-lit-element/

**Comment by thescientist13** [comment]
Got a pretty cool example of a project using Greenwood + Storybook + @web/test-runner working with Lit + CSS Custom properties, building a Single Page Application!  👀 🥳 
https://github.com/AnalogStudiosRI/www.analogstudios.net/pull/4

**Comment by thescientist13** [comment]
One suggestion I've seen to help with the FOUC and WCs is to use [this trick](https://web.dev/declarative-shadow-dom/#fouc)
```css
:not(:defined) {
  opacity: 0;
}
```

To help keep custom elements hidden until their respective `customElements.define` has been called and the browser has loaded it.  Not sure if #576 alleviates this issue much, but still a good technique we can track and explore when the time comes.

**Comment by thescientist13** [comment]
I think with support for Import Attributes now in Greenwood, design system work is a lot more achievable, and so can close this one out.
